chainer/functions/pooling/roi_average_pooling_2d.py

```diff
@@ -0,0 +1,301 @@
+            '''
+            // pos in output filter
+            int pw = i % pooled_width;
+            int ph = (i / pooled_width) % pooled_height;
+            int c = (i / pooled_width / pooled_height) % channels;
+            int n = i / pooled_width / pooled_height / channels;
+
+            int roi_batch_ind = bottom_rois[n * 5 + 0];
+            int roi_start_w = round(bottom_rois[n * 5 + 1] * spatial_scale);
+            int roi_start_h = round(bottom_rois[n * 5 + 2] * spatial_scale);
+            int roi_end_w = round(bottom_rois[n * 5 + 3] * spatial_scale);
+            int roi_end_h = round(bottom_rois[n * 5 + 4] * spatial_scale);
+
+            // Force malformed ROIs to be 1x1
+            int roi_width = max(roi_end_w - roi_start_w + 1, 1);
+            int roi_height = max(roi_end_h - roi_start_h + 1, 1);
+            T bin_size_h = static_cast<T>(roi_height)
+                           / static_cast<T>(pooled_height);
+            T bin_size_w = static_cast<T>(roi_width)
+                           / static_cast<T>(pooled_width);
+
+            int hstart = static_cast<int>(floor(static_cast<T>(ph)
+                                          * bin_size_h));
+            int wstart = static_cast<int>(floor(static_cast<T>(pw)
+                                          * bin_size_w));
+            int hend = static_cast<int>(ceil(static_cast<T>(ph + 1)
+                                        * bin_size_h));
+            int wend = static_cast<int>(ceil(static_cast<T>(pw + 1)
+                                        * bin_size_w));
+
+            // Add roi offsets and clip to input boundaries
+            hstart = min(max(hstart + roi_start_h, 0), height);
+            hend = min(max(hend + roi_start_h, 0), height);
+            wstart = min(max(wstart + roi_start_w, 0), width);
+            wend = min(max(wend + roi_start_w, 0), width);
+            bool is_empty = (hend <= hstart) || (wend <= wstart);
+
+            int bottom_diff_offset =
+                (roi_batch_ind * channels + c) * height * width;
+            int top_offset =
+                (n * channels + c) * pooled_height * pooled_width;
+
+            T count = (hend - hstart) * (wend - wstart);
+            T diff_val = is_empty ? 0. :
+                top_diff[top_offset + ph * pooled_width + pw] / count;
+            for (int h = hstart; h < hend; ++h) {
+                for (int w = wstart; w < wend; ++w) {
+                    int bottom_index = h * width + w;
+                    atomicAdd(
+                        &bottom_diff[bottom_diff_offset + bottom_index],
+                        diff_val);
+                }
+            }
+            ''', 'roi_average_pooling_2d_bwd'
+        )(gy[0], bottom_rois, self.spatial_scale,
+          channels, height, width, self.outh, self.outw,
+          bottom_diff, size=gy[0].size)
+
+        return bottom_diff, None
+
+
+def roi_average_pooling_2d(x, rois, outh, outw, spatial_scale):
```

Comment:
Can we use a pair of ints (a tuple of two ints) for `outh, outw` for consistency with other functions? For example, `unpooling_2d` and `upsampling_2d` have `outsize` parameter which accepts a pair of ints that specifies the output image size (in these functions, `outsize` is optional, so strictly speaking, the requirements of the API design is a bit different, though).